demo-manufacturing/iot-data-intelligence-platform/05-Generative-AI/05.1-ai-tools-iot-turbine-prescriptive-maintenance.py

```
@@ -0,0 +1,607 @@
+principal_id = "" # TO DO: fill with your Service Principal ID
+
+# Create secret scope and put access token as secret value
+w = WorkspaceClient()
+w.secrets.create_scope(scope=secret_scope_name)
+w.secrets.put_secret(scope=secret_scope_name, key=secret_key_name, string_value=secret_value)
+
+# Add permissions on the catalog, schema, VS index and secret to the service principal
+spark.sql(f'GRANT USAGE ON CATALOG {catalog} TO `{principal_id}`')
+spark.sql(f'GRANT USAGE ON SCHEMA {catalog}.{db} TO `{principal_id}`')
+w.grants.update(c.SecurableType.TABLE, vs_index_fullname, 
+                                changes=[c.PermissionsChange(add=[c.Privilege["SELECT"]], principal=principal_id)])
+w.secrets.put_acl(scope=secret_scope_name, principal=principal_id, permission=workspace.AclPermission.READ)
+
+# COMMAND ----------
+
+# MAGIC %md 
+# MAGIC
+# MAGIC Now that our service principal has access to the rights objects, we can create the Vector Search as Tool function below:
+
+# COMMAND ----------
+
+host = w.config.host
+spark.sql(f"""DROP FUNCTION IF EXISTS {catalog}.{db}.turbine_maintenance_reports_retriever_with_secret;""")
+spark.sql(f"""
+CREATE OR REPLACE FUNCTION {catalog}.{db}.turbine_maintenance_reports_retriever_with_secret(search_vector ARRAY<DOUBLE>, databricks_token STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE PYTHON
+AS
+$$
+  try:
+    import requests
+    headers = {{"Authorization": "Bearer "+databricks_token}}
+    #Call our vector search endpoint via simple SQL statement
+    response = requests.post("{host}/api/2.0/vector-search/indexes/{catalog}.{db}.turbine_maintenance_reports_vs_index/query", json ={{"columns":["maintenance_report"],"query_vector":list(search_vector), "num_results":3}}, headers=headers)
+    
+    formatted_response = [item[0] for item in response.json().get('result', {{}}).get('data_array', [])]
+      
+    return formatted_response
+  except Exception as e:
+    return f"Error calling the vs index {{e}}"
```

Not sure, but I think we should raise the error and bubble it up no?